packages/ipc-proxy/src/proxy-generator.ts

```diff
@@ -0,0 +1,214 @@
+        removeIpcListener,
+    };
+};
+
+export const getIpcMethods = (
+    ipcRenderer: IpcRenderer,
+    channelName: string,
+    constructorParams: any,
+) => {
+    const { ipcListeners, addIpcListener, removeIpcListener } = listenersManager();
+
+    const prefix = `${channelName}/${(Date.now() + Math.random()).toString(36)}`;
+
+    let created = false;
+    const init = async () => {
+        if (created) return true;
+        // Handshake with electron main
+        try {
+            await ipcRenderer.invoke(`${channelName}/create`, [prefix, constructorParams]);
+            created = true;
+        } catch (error) {
+            console.error('IpcProxyGenerator', error); // unexpected error
+            throw error;
+        }
+    };
+
+    const addListener = (eventName: string, listener: any) => {
+        const ipcEventName = `${prefix}/event-listener/${eventName}`;
+        addIpcListener(eventName, listener, ipcEventName);
+
+        if (ipcRenderer.listenerCount(ipcEventName) === 0) {
+            // use only one ipc listener and resolve realListeners here
+            ipcRenderer.send(`${prefix}/add-listener`, [eventName, ipcEventName]);
+            ipcRenderer.on(ipcEventName, (_, event) => {
+                ipcListeners
+                    .filter(l => l.ipcEventName === ipcEventName)
+                    .forEach(l => l.realListener(...event));
+            });
+        }
+    };
+
+    const removeListener = (eventName: string, listener: any) => {
+        const ipcEventName = `${prefix}/event-listener/${eventName}`;
+        removeIpcListener(eventName, listener);
+
+        if (ipcListeners.filter(l => l.ipcEventName === ipcEventName).length > 0) {
+            ipcRenderer.removeAllListeners(ipcEventName);
+            ipcRenderer.send(`${prefix}/remove-listener`, [eventName, ipcEventName]);
+        }
+    };
+
+    let requestId = 0;
+
+    const request =
+        (method: string) =>
+        (...args: any[]) =>
+            new Promise<any>((resolve, reject) => {
+                requestId++;
+                const responseEvent = `${prefix}/response/${requestId}`;
+                ipcRenderer.once(responseEvent, (_, response) => {
+                    // success/failure is wrapped in object. see ipcProxyHandler
+                    if (response.success) {
+                        resolve(response.payload);
+                    } else {
+                        reject(response.error);
+                    }
+                });
+                ipcRenderer.send(`${prefix}/request`, [responseEvent, method, args]);
+            });
+
+    const invoke =
+        (method: string) =>
+        (...args: any[]) =>
+            ipcRenderer.invoke(`${prefix}/invoke`, [method, ...args]);
+
+    return {
+        prefix,
+        init,
+        addListener,
+        removeListener,
+        request,
+        invoke,
+    };
+};
+
+const factory = async (
+    ipcRenderer: IpcRenderer,
+    validChannels: string[],
```

Comment:
Why is `validChannels` needed? Only to have all valid channel names together in `preload.ts`?

Reply:
by calling `createProxy('any-channel'...)` in renderer context you whould be able to trigger `ipcRenderer.invoke` with any string on main context. of course handler will not be there so it will hard fail with runtime error.
This is kinda config of accepted ipcRenderer invocations